src/api/libcellml/types.h

```diff
@@ -31,6 +33,13 @@ typedef std::shared_ptr<Units> UnitsPtr; /**< Type definition for shared units p
 class Variable; /**< Forward declaration of Variable class. */
 typedef std::shared_ptr<Variable> VariablePtr; /**< Type definition for shared variable pointer. */
 
+class XmlAttribute; /**< Forward declaration of XmlAttribute class. */
```

Comment:
The main issue with hiding the XML classes and typedefs from the public API is that we still need to handle them from our public-facing `Parser`. I think the best way to get around this is to move the `load...()` methods that handle the XML objects into the parser private implementation. This will mean moving the `load...()` function declarations from the parser header into the `.cpp` file. We'll also need to give the pimpl a pointer to to the parent `Parser` to be able to add errors for the user to recieve. For example (best seen on Github):

```c++
...
/**
 * @brief The Parser::ParserImpl struct.
 *
 * The private implementation for the Parser class.
 */
struct Parser::ParserImpl
{
public:
    Format mFormat;
    Parser *mParser;

    /**
     * @brief Update the @p model with attributes parsed from a @c std::string.
     *
     * Update the @p model with attributes and entities parsed from
     * the @c std::string @p input. Any entities or attributes in @p model with names
     * matching those in @p input will be overwritten.
     *
     * @param model The @c ModelPtr to update.
     * @param input The string to parse and update the @p model with.
     */
    void loadModel(const ModelPtr &model, const std::string &input);

private:
    /**
     * @brief Update the @p component with attributes parsed from @p node.
     *
     * Update the @p component with attributes and entities parsed from
     * the XML @p node. Any entities or attributes in @p component with names
     * matching those in @p node will be overwritten.
     *
     * @param component The @c ComponentPtr to update.
     * @param node The @c XmlNodePtr to parse and update the @p component with.
     */
    void loadComponent(const ComponentPtr &component, const XmlNodePtr &node);
...
};

Parser::Parser(Format format)
    : mPimpl(new ParserImpl())
{
    mPimpl->mFormat = format;
    mPimpl->mParser = this;
}
...
```
The only drawback I could see is that it might arguably look less consistent to have these functions declared/documented outside of the header but as far as I can tell, this [shouldn't matter to doxygen](https://www.stack.nl/~dimitri/doxygen/manual/docblocks.html#cppblock).